File: docs/manual/java/guide/cluster/code/docs/home/persistence/CassandraBlogEventProcessor.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.persistence;

// #imports
import akka.Done;
import com.datastax.driver.core.BoundStatement;
import com.datastax.driver.core.PreparedStatement;
import com.lightbend.lagom.javadsl.persistence.AggregateEventTag;
import com.lightbend.lagom.javadsl.persistence.ReadSideProcessor;
import com.lightbend.lagom.javadsl.persistence.cassandra.CassandraReadSide;
import com.lightbend.lagom.javadsl.persistence.cassandra.CassandraSession;
import org.pcollections.PSequence;

import javax.inject.Inject;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletionStage;

import static com.lightbend.lagom.javadsl.persistence.cassandra.CassandraReadSide.*;
// #imports

public interface CassandraBlogEventProcessor {

  interface Initial {
    // #initial
    public class BlogEventProcessor extends ReadSideProcessor<BlogEvent> {

      private final CassandraSession session;
      private final CassandraReadSide readSide;

      @Inject
      public BlogEventProcessor(CassandraSession session, CassandraReadSide readSide) {
        this.session = session;
        this.readSide = readSide;
      }

      @Override
      public ReadSideProcessor.ReadSideHandler<BlogEvent> buildHandler() {
        // TODO build read side handler
        return null;
      }

      @Override
      public PSequence<AggregateEventTag<BlogEvent>> aggregateTags() {
        // TODO return the tag for the events
        return null;
      }
    }
    // #initial
  }

  public class BlogEventProcessor extends ReadSideProcessor<BlogEvent> {

    private final CassandraSession session;
    private final CassandraReadSide readSide;

    @Inject
    public BlogEventProcessor(CassandraSession session, CassandraReadSide readSide) {
      this.session = session;
      this.readSide = readSide;
    }

    // #tag
    @Override
    public PSequence<AggregateEventTag<BlogEvent>> aggregateTags() {
      return BlogEvent.TAG.allTags();
    }
    // #tag

    // #create-table
    private CompletionStage<Done> createTable() {
      return session.executeCreateTable(
          "CREATE TABLE IF NOT EXISTS blogsummary ( " + "id TEXT, title TEXT, PRIMARY KEY (id))");
    }
    // #create-table

    // #prepare-statements
    private PreparedStatement writeTitle = null; // initialized in prepare

    private CompletionStage<Done> prepareWriteTitle() {
      return session
          .prepare("INSERT INTO blogsummary (id, title) VALUES (?, ?)")
          .thenApply(
              ps -> {
                this.writeTitle = ps;
                return Done.getInstance();
              });
    }
    // #prepare-statements

    // #post-added
    private CompletionStage<List<BoundStatement>> processPostAdded(BlogEvent.PostAdded event) {
      BoundStatement bindWriteTitle = writeTitle.bind();
      bindWriteTitle.setString("id", event.getPostId());
      bindWriteTitle.setString("title", event.getContent().getTitle());
      return completedStatements(Arrays.asList(bindWriteTitle));
    }
    // #post-added

    @Override
    public ReadSideHandler<BlogEvent> buildHandler() {
      // #create-builder
      CassandraReadSide.ReadSideHandlerBuilder<BlogEvent> builder =
          readSide.builder("blogsummaryoffset");
      // #create-builder

      // #register-global-prepare
      builder.setGlobalPrepare(this::createTable);
      // #register-global-prepare

      // #register-prepare
      builder.setPrepare(tag -> prepareWriteTitle());
      // #register-prepare

      // #set-event-handler
      builder.setEventHandler(BlogEvent.PostAdded.class, this::processPostAdded);
      // #set-event-handler

      // #build
      return builder.build();
      // #build
    }
  }
}

File: docs/manual/scala/guide/cluster/code/docs/home/scaladsl/persistence/CassandraBlogEventProcessor.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.scaladsl.persistence

//#imports
import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import akka.Done
import com.datastax.driver.core.BoundStatement
import com.datastax.driver.core.PreparedStatement
import com.lightbend.lagom.scaladsl.persistence.AggregateEventTag
import com.lightbend.lagom.scaladsl.persistence.EventStreamElement
import com.lightbend.lagom.scaladsl.persistence.ReadSideProcessor
import com.lightbend.lagom.scaladsl.persistence.cassandra.CassandraReadSide
import com.lightbend.lagom.scaladsl.persistence.cassandra.CassandraSession
import scala.concurrent.Promise

//#imports

trait CassandraBlogEventProcessor {

  trait Initial {
    //#initial
    class BlogEventProcessor(session: CassandraSession, readSide: CassandraReadSide)(implicit ec: ExecutionContext)
        extends ReadSideProcessor[BlogEvent] {

      override def buildHandler(): ReadSideProcessor.ReadSideHandler[BlogEvent] = {
        // TODO build read side handler
        ???
      }

      override def aggregateTags: Set[AggregateEventTag[BlogEvent]] = {
        // TODO return the tag for the events
        ???
      }
    }
    //#initial
  }

  class BlogEventProcessor(session: CassandraSession, readSide: CassandraReadSide)(implicit ec: ExecutionContext)
      extends ReadSideProcessor[BlogEvent] {

    //#tag
    override def aggregateTags: Set[AggregateEventTag[BlogEvent]] =
      BlogEvent.Tag.allTags
    //#tag

    //#create-table
    private def createTable(): Future[Done] =
      session.executeCreateTable(
        "CREATE TABLE IF NOT EXISTS blogsummary ( " +
          "id TEXT, title TEXT, PRIMARY KEY (id))"
      )
    //#create-table

    //#prepare-statements
    private val writeTitlePromise                     = Promise[PreparedStatement] // initialized in prepare
    private def writeTitle: Future[PreparedStatement] = writeTitlePromise.future

    private def prepareWriteTitle(): Future[Done] = {
      val f = session.prepare("INSERT INTO blogsummary (id, title) VALUES (?, ?)")
      writeTitlePromise.completeWith(f)
      f.map(_ => Done)
    }
    //#prepare-statements

    //#post-added
    private def processPostAdded(eventElement: EventStreamElement[PostAdded]): Future[List[BoundStatement]] = {
      writeTitle.map { ps =>
        val bindWriteTitle = ps.bind()
        bindWriteTitle.setString("id", eventElement.event.postId)
        bindWriteTitle.setString("title", eventElement.event.content.title)
        List(bindWriteTitle)
      }
    }
    //#post-added

    override def buildHandler(): ReadSideProcessor.ReadSideHandler[BlogEvent] = {
      //#create-builder
      val builder = readSide.builder[BlogEvent]("blogsummaryoffset")
      //#create-builder

      //#register-global-prepare
      builder.setGlobalPrepare(() => createTable())
      //#register-global-prepare

      //#register-prepare
      builder.setPrepare(tag => prepareWriteTitle())
      //#register-prepare

      //#set-event-handler
      builder.setEventHandler[PostAdded](processPostAdded)
      //#set-event-handler

      //#build
      builder.build()
      //#build
    }
  }

}
